### PR TITLE
Instance metadata/metadata providers overhaul

### PR DESCRIPTION
**Pull request by JamesAPetts**
See #1460 for motivations.

This PR refactors metadata so that:
- Metadata is indexed in a revamped provider based on its Study, Series and SOPInstanceUIDs.
- Metadata fetched by imageId intelligently decomposes WADO-RS and WADO-URI imageIds to fetch the correct metadata.
- Metadata for the entire header is now stored in naturalized DICOM JSON format. This makes extensions adding support for other modalities or wishing to access modaliy specific or private tags a lot simpler.
- Our metadata provider provides "legacy" tags like the `imagePlaneModule`, `imagePixelModule` etc in the correct format. The 
- Our metadata provider provides a list of (currently only one) fallback tags to use in case of missing tags. (In this example `PixelSpacing` => `ImagerPixelSpacing`.
- JSON launch has a new format that I'll document in an announcement when things are ironed out and this is merged.

The offshoot of this is that we'll stop hitting into issues of "X doesn't work on local, json launched viewer, using WADO-RS whilst using WADO-URI to fetch images, etc." As the metadata fetching is more standardised independent of its source. For instance, one can load MPR on local/json launched now, where as before we saw issues here.

### CODE REVIEW

File: extensions/cornerstone/src/commandsModule.js [hunk 1]
```diff
@@ -248,8 +248,8 @@ const commandsModule = ({ servicesManager }) => {
       if (enabledElement) {
         let viewport = cornerstone.getViewport(enabledElement);
         viewport.voi = {
```
Comment (swederik) [inline extensions/cornerstone/src/commandsModule.js]: Find/replace error I guess. This won't work, cornerstone is expecting lowercase

Comment (JamesAPetts) [reply extensions/cornerstone/src/commandsModule.js]: Oh interesting this one was hit by no tests, fixed the vtkjs one, thanks.

File: platform/core/src/DICOMWeb/getModalities.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
-export default function getModalities(modality, modalitiesInStudy) {
-  if (!modality && !modalitiesInStudy) {
+export default function getModalities(Modality, modalitiesInStudy) {
```
Comment (swederik) [inline platform/core/src/DICOMWeb/getModalities.js]: Might as well change the casing on modalitiesInStudy as well

**Comment by JamesAPetts** [comment]
netlify example broken, need to publish new dcmjs version first!

File: platform/viewer/src/lib/localFileLoaders/dicomFileLoader.js [hunk 3]
```diff
@@ -44,147 +51,56 @@ const DICOMFileLoader = new (class extends FileLoader {
...
       numberOfStudyRelatedInstances: NumberOfStudyRelatedInstances || DICOMWeb.getString(dataset['00201208']),
-      modality: Modality || DICOMWeb.getString(dataset['00080060']),
+      Modality: Modality || DICOMWeb.getString(dataset['00080060']),
       modalitiesInStudy: ModalitiesInStudy || DICOMWeb.getString(dataset['00080061']),
       modalities:
       */
       seriesList: [series],
     };
 
+    console.log(imageId);
```
Comment (swederik) [inline platform/viewer/src/lib/localFileLoaders/dicomFileLoader.js]: You can remove this

File: platform/viewer/src/lib/localFileLoaders/dicomFileLoader.js [hunk 4]
```diff
@@ -44,147 +51,56 @@ const DICOMFileLoader = new (class extends FileLoader {
...
+      StudyInstanceUID: StudyInstanceUID,
+      StudyDate: StudyDate,
+      StudyTime: StudyTime,
+      AccessionNumber: AccessionNumber,
+      ReferringPhysicianName: ReferringPhysicianName,
+      PatientName: PatientName,
+      PatientId: PatientID,
+      PatientBirthDate,
+      PatientSex: PatientSex,
+      StudyId: StudyID,
```
Comment (swederik) [inline platform/viewer/src/lib/localFileLoaders/dicomFileLoader.js]: Casing is off on this and PatientID (and we should use object shorthand)

File: platform/viewer/src/connectedComponents/Viewer.js [hunk 5]
```diff
@@ -216,11 +216,11 @@ class Viewer extends Component {
       });
     }
     if (isStudyLoaded && isStudyLoaded !== prevProps.isStudyLoaded) {
-      const patientId = studies[0] && studies[0].patientId;
+      const PatientId = studies[0] && studies[0].PatientId;
```
Comment (swederik) [inline platform/viewer/src/connectedComponents/Viewer.js]: PatientID would be better, no?

File: platform/viewer/src/components/UserPreferences/HotkeysPreferences.js [hunk 6]
```diff
@@ -70,7 +70,7 @@ const splitHotkeys = hotkeys => {
 
 /**
  * HotkeysPreferences tab
- * It renders all hotkeys displayed into columns/rows
```
Comment (swederik) [inline platform/viewer/src/components/UserPreferences/HotkeysPreferences.js]: unnecessary change

File: platform/viewer/cypress/support/commands.js [hunk 7]
```diff
@@ -368,14 +368,14 @@ Cypress.Commands.add('percyCanvasSnapshot', (name, options = {}) => {
   });
 });
 
-Cypress.Commands.add('setLayout', (columns = 1, rows = 1) => {
+Cypress.Commands.add('setLayout', (Columns = 1, Rows = 1) => {
```
Comment (swederik) [inline platform/viewer/cypress/support/commands.js]: unnecessary change

File: platform/ui/src/elements/form/__docs__/textArea.doc.mdx [hunk 8]
```diff
@@ -16,7 +16,7 @@ import { TextArea } from './../index.js';
   <State
     initial={{
       value: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit...',
-      rows: 10,
```
Comment (swederik) [inline platform/ui/src/elements/form/__docs__/textArea.doc.mdx]: this will probably break the component in the docs

File: platform/ui/src/components/studyList/TablePagination.js [hunk 9]
```diff
@@ -69,7 +69,7 @@ class TablePagination extends PureComponent {
 
   renderRowsPerPageDropdown() {
     return (
-      <div className="form-inline form-group rows-per-page">
+      <div className="form-inline form-group Rows-per-page">
```
Comment (swederik) [inline platform/ui/src/components/studyList/TablePagination.js]: bad find/replace

File: platform/ui/src/components/studyList/TablePagination.js [hunk 10]
```diff
@@ -91,7 +91,7 @@ class TablePagination extends PureComponent {
   render() {
     return (
       <div className="pagination-area">
-        <div className="rows-dropdown">{this.renderRowsPerPageDropdown()}</div>
+        <div className="Rows-dropdown">{this.renderRowsPerPageDropdown()}</div>
```
Comment (swederik) [inline platform/ui/src/components/studyList/TablePagination.js]: bad find/replace

File: docs/latest/extensions/modules/sop-class-handler.md [hunk 11]
```diff
@@ -66,9 +66,9 @@ getSopClassHandlerModule({ servicesManager, commandsManager }) {
         displaySetInstanceUid: 0,
         wadoRoot: study.getData().wadoRoot,
         wadoUri: instance.getData().wadouri,
-        sopInstanceUid: instance.getSOPInstanceUID(),
-        seriesInstanceUid: series.getSeriesInstanceUID(),
-        studyInstanceUid: study.getStudyInstanceUID(),
+        SOPInstanceUID: instance.getSOPInstanceUID(),
+        SeriesInstanceUID: series.getSeriesInstanceUID(),
+        StudyInstanceUID: study.getStudyInstanceUID(),
```
Comment (dannyrb) [inline docs/latest/extensions/modules/sop-class-handler.md]: I understand that we want consistency here; it is a bit odd to have different casing across props for the SOP Class Handler definition.

File: extensions/_example/src/index.js [hunk 12]
```diff
@@ -81,9 +81,9 @@ const sopClassHandlerModule = {
       displaySetInstanceUid: 0, //utils.guid(),
       wadoRoot: study.getData().wadoRoot,
       wadoUri: instance.getData().wadouri,
-      sopInstanceUid: instance.getSOPInstanceUID(),
-      seriesInstanceUid: series.getSeriesInstanceUID(),
-      studyInstanceUid: study.getStudyInstanceUID(),
+      SOPInstanceUID: instance.getSOPInstanceUID(),
+      SeriesInstanceUID: series.getSeriesInstanceUID(),
+      StudyInstanceUID: study.getStudyInstanceUID(),
```
Comment (dannyrb) [inline extensions/_example/src/index.js]: :eyes:

File: extensions/cornerstone/src/OHIFCornerstoneViewport.js [hunk 13]
```diff
@@ -58,24 +49,24 @@ class OHIFCornerstoneViewport extends Component {
...
    * @param {Number} [frameIndex=1]
    * @return {Object} CornerstoneTools Stack
    */
   static getCornerstoneStack(
     studies,
-    studyInstanceUid,
+    StudyInstanceUID,
     displaySetInstanceUid,
-    sopInstanceUid,
+    SOPInstanceUID,
```
Comment (dannyrb) [inline extensions/cornerstone/src/OHIFCornerstoneViewport.js]: Again, a little odd, but I get it. Not sure where we stand on caps

File: docs/latest/extensions/modules/sop-class-handler.md [hunk 14]
```diff
@@ -15,13 +15,13 @@ A `displaySet` has the following shape:
 ```js
 return {
   plugin: 'html',
-  modality: 'SR',
+  Modality: 'SR',
   displaySetInstanceUid: 0,
```
Comment (dannyrb) [inline docs/latest/extensions/modules/sop-class-handler.md]: Consistent UID casing?

File: docs/latest/extensions/modules/sop-class-handler.md [hunk 15]
```diff
@@ -15,13 +15,13 @@ A `displaySet` has the following shape:
...
   wadoRoot: study.getData().wadoRoot,
   wadoUri: instance.getData().wadouri,
-  sopInstanceUid: instance.getSOPInstanceUID(),
-  seriesInstanceUid: series.getSeriesInstanceUID(),
-  studyInstanceUid: study.getStudyInstanceUID(),
+  SOPInstanceUID: instance.getSOPInstanceUID(),
+  SeriesInstanceUID: series.getSeriesInstanceUID(),
+  StudyInstanceUID: study.getStudyInstanceUID(),
```
Comment (dannyrb) [inline docs/latest/extensions/modules/sop-class-handler.md]: ESLint list? That'd be a lot of words to auto-check

File: extensions/cornerstone/src/init.js [hunk 16]
```diff
@@ -3,29 +3,9 @@ import { SimpleDialog } from '@ohif/ui';
 import cornerstone from 'cornerstone-core';
 import csTools from 'cornerstone-tools';
 import merge from 'lodash.merge';
-import queryString from 'query-string';
```
Comment (dannyrb) [inline extensions/cornerstone/src/init.js]: Remove as dependency if now un-used

File: extensions/cornerstone/src/utils/measurementServiceMappings/measurementServiceMappingsFactory.js [hunk 17]
```diff
@@ -82,13 +90,13 @@ const measurementServiceMappingsFactory = measurementService => {
   const _getAttributes = element => {
     const enabledElement = cornerstone.getEnabledElement(element);
     const imageId = enabledElement.image.imageId;
-    const sopInstance = cornerstone.metaData.get('instance', imageId);
-    const sopInstanceUid = sopInstance.sopInstanceUid;
-    const frameOfReferenceUid = sopInstance.frameOfReferenceUID;
-    const series = cornerstone.metaData.get('series', imageId);
-    const seriesInstanceUid = series.seriesInstanceUid;
+    const instance = cornerstone.metaData.get('instance', imageId);
```
Comment (dannyrb) [inline extensions/cornerstone/src/utils/measurementServiceMappings/measurementServiceMappingsFactory.js]: Worth abstracting?

```js
OHIFMetadata.getInstance(imageId);
OHIFMetadata.getTag('tagName', imageId);
```

abstraction?

File: platform/core/src/studies/services/qido/instances.js [hunk 18]
```diff
@@ -44,36 +44,36 @@ function resultDataToStudyMetadata(server, studyInstanceUid, resultData) {
...
       '?requestType=WADO&studyUID=' +
-      studyInstanceUid +
+      StudyInstanceUID +
       '&seriesUID=' +
-      seriesInstanceUid +
+      SeriesInstanceUID +
       '&objectUID=' +
-      sopInstanceUid +
+      SOPInstanceUID +
       '&contentType=application%2Fdicom';
```
Comment (dannyrb) [inline platform/core/src/studies/services/qido/instances.js]: does casing matter here? Either parsing/matching/casing cause issues?

File: platform/i18n/src/locales/en-US/Common.json [hunk 19]
```diff
@@ -8,7 +8,7 @@
   "Play": "Play",
   "Previous": "Previous",
   "Reset": "Reset",
-  "RowsPerPage": "rows per page",
+  "RowsPerPage": "Rows per page",
```
Comment (dannyrb) [inline platform/i18n/src/locales/en-US/Common.json]: :eyes:

File: platform/ui/src/components/quickSwitch/StudiesItem.styl [hunk 20]
```diff
@@ -82,10 +82,10 @@ $spacerY = 12px
     position: absolute
     right: $spacerX
     top: $spacerY
-    .studyDate
+    .StudyDate
       margin-top: 8px
       color: var(--text-secondary-color)
-    .studyDescription
+    .StudyDescription
```
Comment (dannyrb) [inline platform/ui/src/components/quickSwitch/StudiesItem.styl]: Probably don't have to be caps, but it's cool

File: platform/viewer/src/components/Labelling/LabellingFlow.js [hunk 21]
```diff
@@ -151,7 +151,7 @@ const LabellingFlow = ({
         return (
           <SelectTree
             items={OHIFLabellingData}
-            columns={1}
+            Columns={1}
```
Comment (dannyrb) [inline platform/viewer/src/components/Labelling/LabellingFlow.js]: maybe not here?

File: platform/viewer/src/connectedComponents/Viewer.js [hunk 22]
```diff
@@ -136,7 +136,7 @@ class Viewer extends Component {
         timepointType: 'baseline',
         timepointId: 'TimepointId',
         studyInstanceUids: this.props.studyInstanceUids,
-        patientId: filter.patientId,
+        PatientID: filter.PatientID,
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/Viewer.js]: How should we handle this?

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js [hunk 23]
```diff
@@ -91,10 +91,10 @@ const _isQueryParamApplied = (study, filters = {}, isFilterStrategy) => {
     return applied;
   }
 
-  const { seriesList = [], displaySets = [] } = study;
-  const firstSeries = isFilterStrategy ? seriesList[0] : displaySets[0];
+  const { series = [], displaySets = [] } = study;
```
Comment (dannyrb) [inline platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: Do we need to worry about accidental conflicts here?

File: platform/viewer/src/connectedComponents/Viewer.js (hunk 22 above)
Comment (JamesAPetts) [reply platform/viewer/src/connectedComponents/Viewer.js]: We call it with a filter which has a `PatientID` property right now. Its better that its consistent.

File: docs/latest/extensions/modules/sop-class-handler.md (hunk 15 above)
Comment (JamesAPetts) [reply docs/latest/extensions/modules/sop-class-handler.md]: This is a good idea, we should think about it in another issue though.

File: extensions/cornerstone/src/OHIFCornerstoneViewport.js (hunk 13 above)
Comment (dannyrb) [reply extensions/cornerstone/src/OHIFCornerstoneViewport.js]: After reviewing 130 files, I've warmed up to this. It feels odd, but it adds a lot of value I really, really like consistency; now we just need an ESLint rule. Some cases still seem fuzzy to me. Like a PatientName property that is a display representation of the same DICOM tag.

IE. W/ formatting applied, as part of search filter input

We also had a few cases of the same tag within the same lexical scope. This lead to one variable using this official naming format, and the other using a different name (and alternative casing).

File: platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js (hunk 23 above)
Comment (JamesAPetts) [reply platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js]: I've tested the app thoroughly after these changes and haven't encounted any yet.
